Fix "autosave: true" on belongs_to of join model causes invalid records to be saved

lcreid opened the pull request:
### Summary

This PR address issue #28080. The changes to code are to use a variable local to the `save_collection_association` method in `activerecord/lib/active_record/autosave_association.rb`, instead of an instance variable.

Prior to this PR, when there is a circular series of `autosave: true` associations, the callback for a `has_many` association is run while another instance of the same callback on the same association hasn't finished running. When control returns to the first callback, the instance variable has changed, and subsequent associated records aren't saved correctly. Specifically, the ID field for the `belongs_to` corresponding to the `has_many` is nil (despite validations that should prevent this).

The code in this pull request also includes models, table definitions, and test cases.

### Other Information

The comments on issue #28080 show more of the thinking behind this pull request.

lcreid commented:
I fixed the order of the tables I added to `schema.rb` and added `force: :cascade` so the schema would build on RDBMSs other than sqlite. When I run the tests on my local rails-dev-box, I get one failure in the postgresql tests:
```
  1) Failure:
AssociationProxyTest#test_save_on_parent_saves_children [/vagrant/rails/activerecord/test/cases/associations_test.rb:186]:
Expected: 1
  Actual: 2

5437 runs, 15172 assertions, 1 failures, 0 errors, 2 skips
```
Unfortunately, I can only reproduce the failure when I run the entire test suite. Running the individual test case, or the individual file, finds no failures. I will look into this failure further when I get a chance, but that might not be for a few days.

I look forward to your comments on the PR.

lcreid commented:
I'm assuming that this PR is ready to go, since it passes the Travis checks. If you want me to figure out the one issue above, please let me know. Note that I have the same failure if I run the tests on the master branch, which comes straight from the upstream (Rails) repository.

mtsmfm commented:
@lcreid 

> I get one failure in the postgresql tests

It's not related to this PR.
I fixed it on https://github.com/rails/rails/pull/28426 and the fix is merged yesterday.

lcreid commented:
Thanks, @mtsmfm . Should I rebase (is that the right term) my PR? Is there anything I can do to move this PR along?

mtsmfm commented:
> Should I rebase (is that the right term) my PR? Is there anything I can do to move this PR along?

You should rebase and squash your commits :pray: 

Rails' master branch is frequently updated so it's not necessary to keep rebasing every time but in this case it is preferable I think.
However, your should squash your commits into a single commit:
https://github.com/rails/rails/blob/c1317587921688ca081f5c326285ee207fb6e957/guides/source/contributing_to_ruby_on_rails.md#commit-your-changes

lcreid commented:
@chancancode Hi, I'm the old guy you meet at the Vancouver Ruby Meetup a few years ago. You showed me how to fix Rails bugs. I've now submitted a PR for another issue, and would really appreciate your feedback. There are very few changed lines of code, but the implications a potentially large, since they affect autosaving of associations. Thanks in advance.

rafaelfranca commented on activerecord/CHANGELOG.md:
```diff
@@ -1 +1,715 @@
-Please check [5-1-stable](https://github.com/rails/rails/blob/5-1-stable/activerecord/CHANGELOG.md) for previous changes.
+*   Fix circular `autosave: true`
```
Looks like you rebase pulled more lines that we need. Could you keep only the "Please check [5-1-stable](https://github.com/rails/rails/blob/5-1-stable/activerecord/CHANGELOG.md) for previous changes." and you new entry?

lcreid replied:
I believe I've fixed this. I'll push the revised PR shortly.

rafaelfranca commented on activerecord/test/cases/associations/has_many_through_associations_test.rb:
```diff
@@ -1252,4 +1255,56 @@ def test_incorrectly_ordered_through_associations
+    assert_equal [cs180a].sort_by(&:id), fall.sections.sort_by(&:id)
+  end
+
+  def test_belongs_to_autosave_true_fails_for_two_associations
```
Does this test describe a failure? The test content seems like everything work.

lcreid replied:
Do you want me to remove the other tests and leave only `test_belongs_to_autosave_true_fails_for_two_associations`?

gmcgibbon replied:
I would remove the other tests and rewrite this test to be something like this:
```ruby
def test_circular_autosave_association_correctly_saves_multiple_records
  cs180 = Seminar.new(name: "CS180")
  fall = Session.new(name: "Fall")
  sections = [
    cs180.sections.build(short_name: "A"),
    cs180.sections.build(short_name: "B"),
  ]
  fall.sections << sections
  fall.save!
  fall.reload
  assert_equal sections, fall.sections.sort_by(&:id)
end
```

lcreid commented on activerecord/test/cases/associations/has_many_through_associations_test.rb:
```diff
@@ -1252,4 +1255,56 @@ def test_incorrectly_ordered_through_associations
+    assert_equal [cs180a].sort_by(&:id), fall.sections.sort_by(&:id)
+  end
+
+  def test_belongs_to_autosave_true_fails_for_two_associations
```
When I first observed the issue, I noticed that it didn't seem to happen if there was only one element in the association. As I recall, I decided to write a test that would confirm that was the case. So, short answer, `test_belongs_to_autosave_true_works_for_one_association` should pass without the fix included in this PR. `test_belongs_to_autosave_true_fails_for_two_associations` should fail without the fix in this PR.

lcreid commented:
Sorry for the delay in sorting this out. I was near my limit of git-fu. Hopefully this PR is the way you need it now.

lcreid commented:
I get it now. Every time another PR gets merged, mine will have conflicts. Do I have to keep updating my PR to get rid of the conflicts?

lcreid commented:
What can I do to move this PR along? Should I create a new PR for Rails 5.1? Should I just rebase again to resolve the conflicts? Something else? Sorry for the noob questions.

gmcgibbon reviewed (commented):
This is still reproducible in Rails 6 5a71a773eab626319d30f35402b9595a26bfe64e.

gmcgibbon commented:
@lcreid This still seems reproducible on master so I would keep this PR open. Please rebase  and resolve conflicts so your branch is mergeable again. Thanks! 😄

lcreid committed:
Fix circular `autosave: true`

Use a variable local to the `save_collection_association` method in
`activerecord/lib/active_record/autosave_association.rb`, instead of an
instance variable.

Prior to this PR, when there was a circular series of `autosave: true`
associations, the callback for a `has_many` association was run while
another instance of the same callback on the same association hadn't
finished running. When control returned to the first instance of the
callback, the instance variable had changed, and subsequent associated
records weren't saved correctly. Specifically, the ID field for the
`belongs_to` corresponding to the `has_many` was `nil`.

Remove unnecessary test and comments.

Fixes #28080.

lcreid commented:
The Travis build failed in ActiveJob tests, but this PR only changes ActiveRecord. I ran the tests for ActiveJob on my own Vagrant box, and all the tests passed. Any hints as to what I can look at to figure out the Travis failure?

gmcgibbon commented on activerecord/test/models/seminar.rb:
```diff
@@ -0,0 +1,8 @@
+# frozen_string_literal: true
+
+##
+# Represents a seminar.
```
Comments are good 👍 but I don't know if we need these. None of the other test models seem to be commented.

lcreid commented:
@gmcgibbon I think this is shipshape now. Thanks for your feedback and support!
